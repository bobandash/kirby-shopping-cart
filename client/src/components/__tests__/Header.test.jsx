import { describe, it, expect, beforeEach } from "vitest";
import { render, screen } from "@testing-library/react";
import Header1 from "../Header";
import { BrowserRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";

// setups props
describe("Header", () => {
  const mockCartItems = [
    {
      item: "test",
      quantity: 1
    },
    {
      item: "test2",
      quantity: 10
    }
  ];

  describe("DOM", () => {
    it("matches snapshot", () => {
      const routerWrapper = ({ children }) => <BrowserRouter>{children}</BrowserRouter>;
      const { container } = render(<Header1 cartItems={mockCartItems} />, {
        wrapper: routerWrapper
      });
      expect(container).toMatchSnapshot();
    });
  });

  describe("Header Links", () => {
    beforeEach(() => {
      const routerWrapper = ({ children }) => <BrowserRouter>{children}</BrowserRouter>;
      render(<Header1 cartItems={mockCartItems} />, { wrapper: routerWrapper });
    });

    it("kirby logo redirects to home page", async () => {
      const kirbyLogo = screen.getByAltText("Kirby Logo");
      expect(kirbyLogo).toBeInTheDocument();
      const user = userEvent.setup();
      await user.click(kirbyLogo);
      expect(window.location.pathname).toBe("/");
    });

    it("cart icon redirects to cart", async () => {
      const cartIcon = screen.getByAltText("Cart");
      expect(cartIcon).toBeInTheDocument();
      const user = userEvent.setup();
      await user.click(cartIcon);
      expect(window.location.pathname).toBe("/cart");
    });
  });

  describe("cart quantity updates", async () => {
    it("empty cart", () => {
      const emptyCart = [];
      const routerWrapper = ({ children }) => <BrowserRouter>{children}</BrowserRouter>;
      render(<Header1 cartItems={emptyCart} />, { wrapper: routerWrapper });
      const cartIcon = screen.getByAltText("Empty Cart");
      expect(cartIcon).toBeInTheDocument();
    });

    it("less than 99 quantity cart", () => {
      const cartItemsLessThan99 = [
        {
          item: "test",
          quantity: 1
        },
        {
          item: "test2",
          quantity: 10
        }
      ];
      const routerWrapper = ({ children }) => <BrowserRouter>{children}</BrowserRouter>;
      render(<Header1 cartItems={cartItemsLessThan99} />, { wrapper: routerWrapper });
      const cartAmount = screen.getByText(11);
      expect(cartAmount).toBeInTheDocument();
    });

    it("more than 99 quantity cart", () => {
      const cartItemsMoreThan99 = [
        {
          item: "test",
          quantity: 90
        },
        {
          item: "test",
          quantity: 50
        }
      ];
      const routerWrapper = ({ children }) => <BrowserRouter>{children}</BrowserRouter>;
      render(<Header1 cartItems={cartItemsMoreThan99} />, { wrapper: routerWrapper });
      const cartAmount = screen.getByText("99+");
      expect(cartAmount).toBeInTheDocument();
    });
  });
});
