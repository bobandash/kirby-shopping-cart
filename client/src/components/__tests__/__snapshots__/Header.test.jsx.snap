// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header > DOM > matches snapshot 1`] = `
<div>
  <header
    class="_header_1d983c _opaque_1d983c"
  >
    <div
      class="_container-header_1d983c"
    >
      <div
        class="_mobile-icon-container_1d983c"
      >
        <i
          class="fa-2x fa-solid fa-bars"
        />
      </div>
      <div>
        <a
          href="/"
        >
          <img
            alt="Kirby Logo"
            class="_kirby-logo_1d983c"
            src="/src/assets/kirby-logo.png"
          />
        </a>
      </div>
      <form
        class="_search-bar-form_1d983c _desktop-icon-container_1d983c"
      >
        <input
          placeholder="Search Products..."
          type="text"
        />
        <button
          type="submit"
        >
          <i
            class="fa-solid fa-magnifying-glass"
          />
        </button>
      </form>
      <div
        class="_profile-cart-container_1d983c"
      >
        <a
          class="_desktop-icon-container_1d983c"
          href="/"
        >
          <i
            class="fa-2x fa-solid fa-user"
          />
          <span>
            Sign In/Register
          </span>
        </a>
        <a
          href="/cart"
        >
          <div
            class="_cart-icon-container_1d983c"
          >
            <img
              alt="Cart"
              class="_cart-icon_1d983c"
              src="/src/assets/Cart/kirby_in_cart.png"
            />
            <p
              class="_cart-quantity_1d983c _less-than-two-digits_1d983c"
            >
              11
            </p>
          </div>
          <span>
            Cart
          </span>
        </a>
      </div>
    </div>
    <nav
      class="_subheader_8b03fa"
    >
      <div
        class="_container_8b03fa"
      >
        <ul>
          <li>
            <a
              href="/category/all"
            >
              Shop All
            </a>
          </li>
          <li>
            <a
              href="/category/games"
            >
              Games
            </a>
          </li>
          <li>
            <a
              href="/category/plushies"
            >
              Plushies
            </a>
          </li>
          <li>
            <a
              href="/category/keychains"
            >
              Keychains
            </a>
          </li>
        </ul>
      </div>
    </nav>
  </header>
  <aside
    class="_mobile-sidebar_ff7a54 _hidden_ff7a54"
  >
    <ul
      class="_sidebar-links_ff7a54"
    >
      <li
        class="_link_ff7a54 _hidden_ff7a54"
      >
        <a
          href="/category/all"
        >
          Shop All
        </a>
      </li>
      <li
        class="_link_ff7a54 _hidden_ff7a54"
      >
        <a
          href="/category/games"
        >
          Games
        </a>
      </li>
      <li
        class="_link_ff7a54 _hidden_ff7a54"
      >
        <a
          href="/category/plushies"
        >
          Plushies
        </a>
      </li>
      <li
        class="_link_ff7a54 _hidden_ff7a54"
      >
        <a
          href="/category/keychains"
        >
          Keychains
        </a>
      </li>
    </ul>
  </aside>
</div>
`;
